Modernization of the KerasCV Object Detection guide

LukeWood opened the pull request:
This is a huge overhaul to the KerasCV OD API!

Before publishing I have a few action items:

- [x] fix numerical differences between PyCOCOCallback and BoxCOCOMetrics (should be quick)
- [x] update the visualization utility to make text not overflow image
- [x] allow customization of font size in visualization utility
- [x] show confidence scores in the visualization utility

----

Demo colab: https://colab.sandbox.google.com/github/LukeWood/keras-io/blob/odguidev2/guides/ipynb/keras_cv/object_detection_keras_cv.ipynb

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,750 @@
+"""
+Next, lets batch our data.
+
+In KerasCV object detection tasks it is recommended that
```
Why do the batches themselves need to be ragged? That doesn't seem to be related to image size of having a different number of labels in a dict.

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,750 @@
+
+## Data augmentation
+
+One of the most labor-intensive tasks when constructing object detection
```
s/labor-intensive/challenging

We're all having fun, right?

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,750 @@
+"""
+
+model = keras_cv.models.RetinaNet.from_preset(
+    "resnet50_v2_imagenet",
```
s/resnet50_v2_imagenet/resnet50_imagenet

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,750 @@
+
+Next, let's construct a learning rate schedule and optimizer for our model.
+"""
+base_lr = 0.005 * (BATCH_SIZE / 16)
```
Might be good to offer some more explanation and visualize the LR schedule. This is an advanced topic.

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,750 @@
+"""
+base_lr = 0.005 * (BATCH_SIZE / 16)
+lr_decay = tf.keras.optimizers.schedules.PiecewiseConstantDecay(
+    boundaries=[12000 * 16, 16000 * 16],
```
The `boundaries` arg seems arbitrary. I would offer intuition and some rules of thumb. Why 16? Why 12K?

LukeWood replied:
My honest answer is that anecdotally I found these to do well.

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,750 @@
+
+And important but easy to miss step in training your own object detection model
+is freezing the `BatchNormalization` layers in your backbone.
+We felt it was a violation of user expectations to automatically do this,
```
We felt it would be confusing to do this automatically

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,750 @@
+To force an evaluation, you may call `coco_metrics.result(force=True)`:
+"""
+model.compile(
+    classification_loss="focal",
```
These loss functions deserve some (even brief) explanation since they are custom to this problem.

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,750 @@
+
+In this section, we will use a `keras_cv` provided preset:
+"""
+model = keras_cv.models.RetinaNet.from_preset(
```
Didn't we already define this on line 58?

LukeWood replied:
Yes, but by doing this we re-initialize the model to make sure we have a fully pretrained model.  Above we override it, and then we don't get strong convergence necessarily.

jbischof replied:
Should we call it `pretrained_model`?

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,750 @@
+And run `model.fit()`!
+"""
+
+# If you want to train the fully model, uncomment `.take(20)` from each
```
I would suggest showing how to instantiate a model from a pretrained backbone and then running a small number of steps without worrying about visualization and metric values. This is what most users will want to do and we want a "plug and play" example for them to copy.

LukeWood replied:
Last time we shipped the guide it felt pretty incomplete without a dedicated inference section.  

Personally, I say we keep it!

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,750 @@
+- tune the hyperparameters and data augmentation used to produce high quality results
+- train an object detection model on your own dataset
+
+One last fun code snippet to showcase the power of KerasCV's modular API!
```
I'm not sure this showcases modularity. We're passing the outputs of one monolithic API to another. It does showcase the diversity of our offerings however.

LukeWood replied:
removed -modularity

jbischof reviewed (commented):
Amazing start @LukeWood! Starting with some high level suggestions.

I think the next steps are to rearrange this guide as follows:
1. High level explanation of OD problem
2. Understanding the input data (both tensors and image visualizaton)
3. Understanding the metrics
4. Load a pretrained model, evaluate the metrics, and visualize some of the predictions
5. Show how to instantiate a model with pretrained backbone only
6. Explain how to set up the learning rate and loss functions
7. Run a couple of training steps for demonstration purposes and give some advice on how long the model would actually need to train. Do not bother with the metrics or predictions (they will be bad).

Finally, let's keep the tone light and helpful. This is supposed to be fun!

LukeWood commented on guides/keras_cv/object_detection_keras_cv.py:
````diff
@@ -0,0 +1,750 @@
+```
+
+Clearly yields bounding boxes in the format `xywh`.  You can read more about
+KerasCV bounding box formats [in the API docs](https://keras.io/api/keras_cv/bounding_box/formats/).
````
Good catch, I've moved this higher up.

LukeWood commented:
Ok @jbischof @fchollet @ianstenbit ready for another round of reviews!

LukeWood commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,750 @@
+resource.setrlimit(resource.RLIMIT_NOFILE, (high, high))
+
+"""
+## Data loading
```
Reorganized a bit, let me know what you think now

jbischof replied:
Looks great thanks!

fchollet commented on guides/keras_cv/object_detection_keras_cv.py:
````diff
@@ -0,0 +1,868 @@
+
+The data often looks something like this:
+```
+image = [height, width, 3])
````
Remove extra parens

fchollet commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,868 @@
+
+
+This is done because there is no one correct format for bounding boxes!
+Components in different pipelines expect different formats, and so by requiring them to be specified we ensure that our components remain readable, reusable, and clear.
```
Break long line

fchollet commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,868 @@
+Our data pipeline is now complete!
+We can now move on to model creation and training.
+
+### Model creation
```
Should be a lvl-2 section (or create an encompassing lvl-2 section for "model architecture", "optimizer", "loss functions", "metrics")

jbischof replied:
I think the untrained model creation should occur after you discuss loss and metrics on the pretrained model

fchollet commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,868 @@
+            # passing 1e9 ensures we never evaluate until
+            # `metrics.result(force=True)` is
+            # called.
+            evaluate_freq=1e9,
```
Is this really necessary? Why not just remove the call to `result()` in `on_epoch_end`?

LukeWood replied:
It just helps performance to make it 1e9, we only call result() once in this callback per epoch.  Thats the only place we read the metrics from.

fchollet commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,868 @@
+- tune the hyperparameters and data augmentation used to produce high quality results
+- train an object detection model on your own dataset
+
+One last fun code snippet to showcase the power of KerasCV's API!
```
Nice touch!

fchollet commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,868 @@
+"""
+Object detection, using a pretrained ML model, on ML generated images, all using
+KerasCV models!
+Awesome!
```
You can drop that last one

fchollet reviewed (commented):
Thanks for the PR!

Since the guide URL is being changed, please add an entry in `redirects/` to redirect the old location to the old one.

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,868 @@
+"""
+
+def dict_to_tuple(inputs):
+    return inputs["images"], bounding_box.to_dense(
```
Is this a static method on a class `bounding_box`? Is `bounding_box` just a namespace? Hard to tell

LukeWood replied:
bounding box is a namespace!

jbischof replied:
Never would have guessed!

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,868 @@
+### Model creation
+
+We'll use the KerasCV API to construct a RetinaNet model.
+In this tutorial we use a pretrained ResNet50 backbone, initializing the
```
you can just say "using pretrained weights produced"

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,868 @@
+
+model = keras_cv.models.RetinaNet.from_preset(
+    "resnet50_imagenet",
+    # number of classes to be used in box classification
```
This is probably a good opportunity to explain the `+ 1`

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,868 @@
+so instead we recommend you do it in your training loops.
+"""
+
+for layer in model.backbone.layers:
```
Should this be a method on `RetinaNet`? Seems very handy

```python
model.freeze_batchnorm_layers()
model.unfreeze_batchfrom_layers()
```

LukeWood replied:
maybe in a follow up, for now lets leave it.

BatchNorm is something I think Keras in general could handle a bit better - (sync batch norm, normal batch norm, freezing, unfreezing)

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,868 @@
+
+### Optimizer
+
+In this guide, we use a standard SGD optimizer and rely on the [`keras.callbacks.ReduceLROnPlateau`](https://keras.io/api/callbacks/reduce_lr_on_plateau/) callback to reduce the learning rate.
```
These are really long lines.

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,868 @@
+In KerasCV you can use these losses simply by passing the strings `"focal"` and `"smoothl1"` to `compile()`:
+"""
+
+pretrained_model.compile(
```
`model.compile(...)`?

LukeWood replied:
Now that I moved this model doesn't exist anymore at this cell 👍

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,868 @@
+)
+
+"""
+If you want to train the fully model, uncomment `.take(20)` from each
```
Thank you, this is great!

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,868 @@
+        return logs
+
+"""
+## Training our model
```
I think this is the penultimate section of the guide. It's weird to go back and forth between the pretrained and random models.

LukeWood replied:
I see what you are saying here, but I was struggling to come up with a better solution.

jbischof reviewed (commented):
This is SO much better, thank you!

My biggest remaining suggestion is to essentially do everything with the pretrained model FIRST and only at the end show the extra steps needed to train one from scratch. It's fairly confusing to keep going back and forth between the pretrained and random models.

I have trouble reviewing the raw python personally because I can't see any of the images produced by the runtime. Would love to see a notebook version.

LukeWood commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,868 @@
+
+This argument describes *exactly* what format the values in the `"boxes"`
+field of the label dictionary take in your pipeline.
+For example, a box in `xywh` format at coordinates (100, 100) with a width of 55 and a height of 70 would be represented by:
```
Upper left for XY

LukeWood replied:
center_xy is middle

LukeWood commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,868 @@
+the input's aspect ratio, the model will perform signficantly poorer.  For the
+pretrained `"retinanet_resnet50_pascalvoc"` preset we are using, the final
+`MeanAveragePrecision` on the `pascalvoc/2012` evaluation set drops to `0.15`
+from `0.33` when using a naive resizing operation.
```
Removed a sentence and think its a bit clearer now

jbischof replied:
It's still not clear to me why it has to be divisible by 64. Sorry if I'm being dense.

LukeWood commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,868 @@
+Whether you're an object detection amateur or a well seasoned veteran, assembling
+an object detection pipeline from scratch is a massive undertaking.
+Luckily, all KerasCV object detection APIs are built as modular components.
+Whether you need a complete pipeline, an object detection pipeline, or even just
```
Meant to say mode, thanks!

LukeWood commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,868 @@
+configuration parameters.
+"""
+
+BATCH_SIZE = 4
```
Ah, sorry this was needed to open multiple TFDS datasets.  For just 2007 its not needed

LukeWood commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,868 @@
+## Data loading
+
+To get started, let's discuss data loading and bounding box formatting.
+KerasCV has a predefined specificication for bounding boxes.
```
or just "format". updated

LukeWood commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,868 @@
+The code below loads the Pascal VOC dataset, and performs on-the-fly bounding box
+friendly data augmentation inside of a `tf.data` pipeline.
+"""
+augmenter = keras.Sequential(
```
Actually yes

LukeWood commented:
Ok @jbischof @fchollet @ianstenbit ready for another round!

LukeWood commented:
> This is SO much better, thank you!
> 
> My biggest remaining suggestion is to essentially do everything with the pretrained model FIRST and only at the end show the extra steps needed to train one from scratch. It's fairly confusing to keep going back and forth between the pretrained and random models.
> 
> I have trouble reviewing the raw python personally because I can't see any of the images produced by the runtime. Would love to see a notebook version.

I will render one for you now.

LukeWood commented:
The visual results somehow got worse on my last training run despite the MaP remaining 0.33, I am going to try again w/o freezing batchnorm.  🤔

LukeWood commented:
I removed the tuning your prediction decoding section, because I realized my approach was nonsense.  You will NEVER get worse results by allowing your model to make more predictions if you sort by confidence score, so I was basically just selecting the most "flexible" parameterization.  Updated now

LukeWood commented:
Ok, @jbischof and @fchollet I think its finally ready!!!!!

LukeWood commented:
@jbischof to see the preview, please visit

https://colab.sandbox.google.com/github/LukeWood/keras-io/blob/odguidev2/guides/ipynb/keras_cv/object_detection_keras_cv.ipynb

You will need to swap out the pip install for:
```
!pip install --upgrade git+https://github.com/keras-team/keras-cv@rnv2
```

LukeWood commented:
Ok, need to re-train the model one more time with the softmax->sigmoid fix!

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,859 @@
+
+"""
+In order to support easy this easy and intuitive inference workflow, KerasCV
+perform non-max suppression inside of the `RetinaNet` class.
```
s/perform/performs

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,846 @@
+    split="train", dataset="voc/2007", bounding_box_format="xywh"
+)
+eval_ds = load_pascal_voc(split="test", dataset="voc/2007", bounding_box_format="xywh")
+eval_ds = load_pascal_voc(dataset="voc/2007", split="test", bounding_box_format="xywh")
```
You have this line twice

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,846 @@
+Let's make sure our dataset is following the format KerasCV expects.
+By using the `visualize_dataset()` function, you can visually verify
+that your data is in the format that KerasCV expects.  If the bounding boxes
+are not visible, or are visible in the wrong locations that is a sign that your
```
delete comma

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,846 @@
+
+"""
+Due to the fact that the resize operation differs between the train dataset,
+which uses `JitteredResize()` to resize images, and the inference dataset which
```
comma after "inference dataset".

You should likely start a new sentence with "It is good practice..."

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,846 @@
+)
+
+"""
+Great!  We now have a bounding box friendly data augmentation pipeline.
```
When multiple words create a single adjective you should add dashes for clarity:
* state-of-the-art
* out-of-the-box

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,846 @@
+visualize both datasets:
+"""
+
+visualize_dataset(
```
I don't see any outputs in the notebook for this

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,846 @@
+)
+
+"""
+Finally, let's unpackage our inputs from the preprocessing dictionary, and
```
Does this mean that the notebook is not GPU compatible by default?

LukeWood replied:
No it is GPU compatible either way

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,846 @@
+    seed=1231,
+)
+y_pred = model.predict(images)
+visualization.plot_bounding_box_gallery(
```
Still don't see output

jbischof reviewed (approved):
Looks great! Really high quality.

The biggest issue remaining is that I don't see output for a lot of print and visualization statements. I hope that can be remedied before pushing. The rest is mainly grammatical.

LukeWood commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,846 @@
+result = pretrained_model.evaluate(eval_ds.take(40), verbose=0)
+result = coco_metrics.result(force=True)
+
+print_metrics(result)
```
Notebooks don't include output on Keras.io (confirmed by @fchollet ).

this is done to simplify the process of rendering/including them on the website.

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,848 @@
+)
+
+"""
+That looks a lot better!
```
This is quite confusing because NMS is set up as *suppressing* noisy detections (false positives) and instead it is recovering true positives. Is this a good example?

LukeWood replied:
Updated the picture to reference the doc picture instead!  Thanks!

jbischof commented on guides/img/object_detection_keras_cv/object_detection_keras_cv_16_1.png:
Shouldn't this file be out?

LukeWood replied:
out of the guide?  removed?  This is still shown before customizing the decoder.  I feel we should update the default decoder now that my softmax->sigmoid fix is in.

jbischof commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,848 @@
+Non-max suppression is a traditional computing algorithm that solves the problem
+of a model detecting multiple boxes for the same object.
+
+**Example of pre-NMS detections:**
```
Couldn't this come from the API itself? We just have to overwrite the prediction decoder twice, right?

LukeWood replied:
it could, but I feel its weird to write to it for that demonstration purpose.  I wrote it out and it just made the guide look a bit strange.   Also, this isnt a truly non-nmsed image, its just a "barely nmsed" image.  Feel its burdening the user with maybe more logic than they need

LukeWood replied:
actually, this is better !  Updated

fchollet reviewed (approved):
LGTM

martin-gorner commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,872 @@
+def visualize_dataset(inputs, value_range, rows, cols, bounding_box_format):
+    inputs = next(iter(inputs.take(1)))
+    images, bounding_boxes = inputs["images"], inputs["bounding_boxes"]
+    visualization.plot_bounding_box_gallery(
```
Ideally, you should not have to create a visualize_dataset function. This should be enough:
plot_bounding_box_gallery(ds.get_sigle_element(), value_range=(0,255), bounding_box_format="xywh", class_mapping=...)

martin-gorner commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,872 @@
+"""
+
+base_lr = 0.005
+# including a global_clipnorm is extremely important in object detection tasks
```
In my experience, a global clipnorm will prevent training runs from going to NaNs, but they will still be bad. This will show as bumpy training curves. Essentially, the clip norm puts the training back on track when it diverges but the training epochs up to that point are lost. Once you have found correct LR settings, global clipnorm is no longer necessary. You can look at all the "arthropods" samples, even the 32-core TPU pod one using a batch size of 1024. They all converge nicely and none use global clipnorm.

martin-gorner commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,872 @@
+### Optimizer
+
+In this guide, we use a standard SGD optimizer and rely on the
+[`keras.callbacks.ReduceLROnPlateau`](https://keras.io/api/callbacks/reduce_lr_on_plateau/)
```
There is no ReduceLROnPlateau in your .fit() !

martin-gorner commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,872 @@
+Due to the high computational cost of computing COCO metrics, the KerasCV
+`BoxCOCOMetrics` component requires an `evaluate_freq` parameter to be passed to
+its constructor.  Every `evaluate_freq`-th call to `update_state()`, the metric
+will recompute the result.  In between invocations, a cached version of the
```
"cached version": essentially, the metric is returning a wrong value most of the time. This is really not the ideal solution. Please implement a proper "eval-every-N-epochs" functionality instead.

martin-gorner commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,872 @@
+"""
+
+
+class EvaluateCOCOMetricsCallback(keras.callbacks.Callback):
```
This is pretty bad. It means that all KerasCV OD examples will be TPU-incompatible out of the box.

martin-gorner commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,872 @@
+`"smoothl1"` to `compile()`:
+"""
+
+pretrained_model.compile(
```
you are compiling your model twice before training it. Consider moving this "losses" block after the instantiation of the model, where a .compile step is relevant.

martin-gorner commented on guides/keras_cv/object_detection_keras_cv.py:
```diff
@@ -0,0 +1,872 @@
+## Inference and plotting results
+
+KerasCV makes object detection inference simple.  `model.predict(images)`
+returns a RaggedTensor of bounding boxes.  By default, `RetinaNet.predict()`
```
can you be more explicit about the retuned format here ? There are classes too aren't there ?

martin-gorner reviewed (commented):
This is a great tutorial. Thank you Luke for writing it!

Top comments (P0):
- Soft NMS is a top priority in object Detection
- There should be ONE format for images with bounding boxes and make sure everything in KerasCV uses it and is compatible: preprocessing layers, models, visualization utilities etc.
- JitteredResize needs renaming and width and height arguments before release to align it on other augmentation layers.
- remove the evaluate_freq arg from BoxCOCOMetric which essentially has the metric return wrong values, and implement a proper "ecval-every-N-epochs" functionality. (evaluate_freq can be a short-term workaround if needed)

Some other important stuff (P1):
- TPU compatibility ? Can we make it so that people do not have to write their own custom callback ?
- include_rescaling missing in RetinaNet instantiation ?
- educate people about the rampup-sustain-decay LR schedule for fine-tuning OD

One additional change that is cosmetic but will remove many lines that have no educational value: make sure plot_bounding_box_gallery does not need the scale, font_scale, rows and cols arguments out of the box.

LukeWood commented:
> This is a great tutorial. Thank you Luke for writing it!
> 
> Top comments (P0):
> 
> * Soft NMS is a top priority in object Detection
> * There should be ONE format for images with bounding boxes and make sure everything in KerasCV uses it and is compatible: preprocessing layers, models, visualization utilities etc.
> * JitteredResize needs renaming and width and height arguments before release to align it on other augmentation layers.
> * remove the evaluate_freq arg from BoxCOCOMetric which essentially has the metric return wrong values, and implement a proper "ecval-every-N-epochs" functionality. (evaluate_freq can be a short-term workaround if needed)
> 
> Some other important stuff (P1):
> 
> * TPU compatibility ? Can we make it so that people do not have to write their own custom callback ?
> * include_rescaling missing in RetinaNet instantiation ?
> * educate people about the rampup-sustain-decay LR schedule for fine-tuning OD
> 
> One additional change that is cosmetic but will remove many lines that have no educational value: make sure plot_bounding_box_gallery does not need the scale, font_scale, rows and cols arguments out of the box.

Thank you!  I will create a follow up PR for these.

innat-asj commented:
Hi, @LukeWood thanks for making such wonderful instructions. I have a couple of opinions that you might want to consider in the follow-up PR. 

1. The title (**Object Detection with KerasCV**) should be renamed with **2D Object ..**, if 3D OD should be in separated docs. 
2. `.from_pretset` API might be a new look to Keras users. Also, how many backbones are supported currently, and how to approach adding a custom backbone (i.e. output of backbone format) might be good if discussed in the doc.
3. Please use mixed precision, jit enabled, and GPU-memory growth at the beginning to showcase the feature complete document and encourage best practices. 
4. A complete TPU-compatible demonstration would be game-changing. This part should be emphasized more, (IMHO).
5. A demonstration of a custom data loader would be nice with tf.data API.
6. Augmentation like CutMix, MixUp, and RandAugment are mentioned in the description but aren't demonstrated in the code. Adding, if they can be part of the model to get GPU acceleration, no!
7. The parameter of the `model. compile` may need some discussion, i.e. loss, classification_loss, box_loss, etc, lots of params!
8. ..Would be nice if coco_metrics output can be formatted in a table in mode. fit log, no?
10. FasterRCNN should be discussed too as it's present at the time of writing the doc.
11. Saving the trained OD model with the latest format (`.keras`) and reloading successfully. (pls check https://github.com/keras-team/keras-io/issues/1235) 
12. Performance (speed/accuracy) compared to existing models!

---

For TPU, it's written as follows, where the same validation data set (`eval_ds` is passed to the model again via callback, even though it's already processed via `validation_data`). It reminds me of this feature request https://github.com/keras-team/keras/issues/17350 .
cc. @rchao 

```
model.fit(
    train_ds.take(20),
    validation_data=eval_ds.take(20),
    # Run for 10-35~ epochs to achieve good scores.
    epochs=1,
    callbacks=[EvaluateCOCOMetricsCallback(eval_ds.take(20))],
)
```

jbischof commented:
Yes @innat eventually we will write a "Getting Started" guide for KerasCV. We're waiting for a variety of tasks to be ready so that it is more than a few lines 😄

jbischof commented:
I will add @LukeWood that I still find that [image](https://keras.io/img/guides/object_detection_keras_cv/object_detection_keras_cv_16_1.png) with NMS set too high pretty confusing. I would just remove

innat-asj commented:
@LukeWood 

Thanks for the prompt response. 

> Please use mixed precision, jit enabled, and GPU-memory growth at the beginning to showcase the feature complete document and encourage best practices.
Thanks! Can you file a bug on kerasCV to verify this behavior for RetinaNet?

I tried mixed precision, the `keras_cv.layers.Resizing` caused some issues; but `JirtterResize` passed. Maybe bug!

```
filepath = tf.keras.utils.get_file(origin="https://i.imgur.com/gCNcJJI.jpg")
image = keras.utils.load_img(filepath)
image = np.array(image)

inference_resizing = keras_cv.layers.Resizing(
    640, 640, 
    pad_to_aspect_ratio=True, 
    bounding_box_format="xywh"
)

jittered_resize = keras_cv.layers.JitteredResize(
    target_size=(640, 640),
    scale_factor=(0.8, 1.25),
    bounding_box_format="xywh",
    )

keras_resize = keras.layers.Resizing(
    640, 640, crop_to_aspect_ratio=False
)

image_batch = jittered_resize([image]) # OK
image_batch = keras_resize([image]) # OK 
image_batch = inference_resizing([image]) # Error

InvalidArgumentError: Exception encountered when calling layer 'resizing' (type Resizing).

TensorArray dtype is float16 but Op is trying to write dtype float32 
guide
Call arguments received by layer 'resizing' (type Resizing):
  • inputs=['tf.Tensor(shape=(3024, 4032, 3), dtype=uint8)']
```

> Perhaps this is deserving of its own guide? The guide is TPU-friendly, so there's really no delta needed here. What would you like to see to show this off?

Agree, this might need a separate guide (maybe like [mult-device](https://keras.io/guides/distributed_training/) training/inference doc). Sometimes it happens that a well-optimized code perfectly run on GPU (s) but fails to run on TPU (hard experience from Kaggle -_-). That's why I was suggesting if there's a full-proof demonstration that would be really great (discussing possible issues and workaround i.e. like you did `to_ragged() to to_dense()`).

> For TPU, it's written ..
model.fit(
    train_ds.take(20),
    validation_data=eval_ds.take(20),
    # Run for 10-35~ epochs to achieve good scores.
    epochs=1,
    callbacks=[EvaluateCOCOMetricsCallback(eval_ds.take(20))],
)
Not sure how to fix this last one.

Understood. Right now the above way is the only standard way. This might become too heavy if `eval_ds` is large and `model` is big enough. I needed something similar which lead to this FR https://github.com/keras-team/keras/issues/17350.
